lemminx-liberty/src/main/java/io/openliberty/tools/langserver/lemminx/services/FeatureService.java

```diff
@@ -529,4 +534,149 @@ public List<Feature> readFeaturesFromFeatureListFile(List<Feature> installedFeat
+     * @param documentURI    xml document
+     * @return platform list
+     */
+    public Set<String> getAllPlatformsForVersionLessFeature(String featureName, String libertyVersion, String libertyRuntime, int requestDelay, String documentURI) {
+        Optional<Feature> feature = this.getFeature(featureName, libertyVersion, libertyRuntime, requestDelay, documentURI);
+        Set<String> featureNames = new HashSet<>();
+        if (feature.isPresent()) {
+            addRequiredFeatureNames(feature.get(), featureNames);
+            addRequireTolerateFeatureNames(feature.get(), featureNames);
+        }
+        return featureNames.stream()
+                .map(f -> getAllPlatformsForFeature(f, libertyVersion, libertyRuntime, requestDelay, documentURI))
+                .flatMap(Set::stream).collect(Collectors.toSet());
+    }
+
+    /**
+     * find require to tolerate feature names. find all versions
+     * @param feature current feature
+     * @param featureNames feature name list
+     */
+    private static void addRequireTolerateFeatureNames(Feature feature, Set<String> featureNames) {
+        List<FeatureTolerate> featureTolerates = feature.getWlpInformation().getRequireFeatureWithTolerates();
+
+        for(FeatureTolerate featureTolerate: featureTolerates) {
+            String extractedFeatureName = featureTolerate.getFeature().contains(IO_OPENLIBERTY_INTERNAL_VERSIONLESS) ?
+                    featureTolerate.getFeature().substring(
+                            featureTolerate.getFeature()
+                                    .lastIndexOf(IO_OPENLIBERTY_INTERNAL_VERSIONLESS) + IO_OPENLIBERTY_INTERNAL_VERSIONLESS.length())
+                    : featureTolerate.getFeature();
+            String extractedFeatureNameWithoutVersion = extractedFeatureName.contains("-") ?
+                    extractedFeatureName.substring(0,extractedFeatureName.lastIndexOf("-") + 1) :
+                    extractedFeatureName;
+            featureNames.add(extractedFeatureName);
+            if(featureTolerate.getTolerates()!=null && !featureTolerate.getTolerates().isEmpty()){
+                for (String tolerateVersion:featureTolerate.getTolerates()){
+                    featureNames.add(extractedFeatureNameWithoutVersion + tolerateVersion);
+                    // if feature names are changed recently. like ejb-3.2 to enterprisebeans-4.0
+                    if (LibertyConstants.changedFeatureNameMap.containsKey(extractedFeatureNameWithoutVersion)) {
+                        featureNames.add(LibertyConstants.changedFeatureNameMap.get(extractedFeatureNameWithoutVersion) + tolerateVersion);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * add all required feature names
+     * @param feature current feature
+     * @param requiredFeatureNames required feature name array
+     */
+    private static void addRequiredFeatureNames(Feature feature, Set<String> requiredFeatureNames) {
```

Comment:
I don't think there is a reason for this method to be static.

Reply:
Corrected